src/Experimental/degeneracy.jl

```diff
@@ -0,0 +1,113 @@
+
+Return the core number for each vertex in graph `g`.
+
+### References
+Parallel k-Core Decomposition on Multicore Platforms, Humayun Kabir and Kamesh Madduri, 2017.
+https://doi.org/10.1109/IPDPSW.2017.151
+
+### Performance
+Time complexity is ``\\mathcal{O}(K_{max}*|V| + |E|)``.
+"""
+function PKC(g::AbstractSimpleGraph{T}, frac = 0.95) where T
+    has_self_loops(g) && throw(ArgumentError("graph must not have self-loops"))
+    deg = Atomic{T}.(degree(g))
+    buff = [Vector{T}() for _ in 1:nthreads()]
+    buff_end = zeros(Int64, nthreads())
+    visited = 0
+    level = 0
+    while visited < nv(g)*frac
+        visited += process_level(g, deg, level, buff, buff_end)
+        level += 1
+    end
+    if visited < nv(g)
+        g_small, vmap = subgraph(g, deg, level, nv(g)-visited)
+        U = eltype(g_small)
+        newdeg = Atomic{U}.(degree(g_small))
+        newbuff = [Vector{U}() for _ in 1:nthreads()]
+        while visited < nv(g)
+            visited += process_level(g_small, newdeg, level, newbuff, buff_end)
+            level += 1
+        end
+        @threads for i in 1:nv(g_small)
+            deg[vmap[i]][] = newdeg[i][]
+        end
+    end
+    core_num = Vector{T}(undef, nv(g))
+    @threads for v in 1:nv(g)
+        core_num[v] = deg[v][]
+    end
+    return core_num
+end
+
+function process_level(g::AbstractGraph{T}, deg::Vector{Atomic{T}}, level::Int64,
+                       buff::Vector{Vector{T}}, buff_end::Vector{Int64}) where T
+    buff_end .= 0
+    @threads for v in 1:nv(g)
+        if deg[v][] == level
+            tid = threadid()
+            push!(buff[tid], v)
+            buff_end[tid] += 1
+        end
+    end
+    @threads for tid in 1:nthreads()
+        while !isempty(buff[tid])
+            v = popfirst!(buff[tid])
+            for u in all_neighbors(g, v)
+                if deg[u][] > level
+                    a = atomic_sub!(deg[u], one(T))
+                    if a == level+1
+                        push!(buff[tid], u)
+                        buff_end[tid] += 1
+                    end
+                    if a <= level
+                        atomic_add!(deg[u], one(T))
+                    end
+                end
+            end
+        end
+    end
+    return sum(buff_end)
+end
+
+function subgraph(g::AbstractSimpleGraph{T}, deg::Vector{Atomic{T}}, level::Int64,
+                  nvg_small::Int64) where T
+    U = UInt8
+    while nvg_small > typemax(U)
+        U = widen(U)
+    end
+    g_small = SmallGraph{U}(nvg_small)
+    in_gsmall = falses(nv(g))
+    @threads for v in 1:nv(g)
+        if deg[v][] >= level
+            in_gsmall[v] = true
+        end
+    end
+    vmap = findall(in_gsmall)
+    newvid = Vector{Int64}(undef, nv(g))
+    @threads for i in 1:length(vmap)
+        newvid[vmap[i]] = i
+    end
+    @threads for s in vmap
+        for d in all_neighbors(g, s)
+            if in_gsmall[d]
+                add_edge!(g_small, newvid[s], newvid[d])
```

Comment:
sorry - how are you guaranteeing that this `add_edge!` is 1) not duplicating edges, and 2) inserting in sorted order?